### PR TITLE
[Perf][CI] reduce gpu-smoke smoke scope for tests/ops

### LINKED ISSUE #736: [PERF][CI] reduce gpu-smoke smoke scope for the pr critical path
### Symptom / Motivation
`gpu-smoke.yml` currently spends too much time inside `Run tests`, which is the largest stable contributor to `gpu-smoke` job latency. The performance analysis identifies this as the P0 optimization target for the PR critical path.

For PR-triggered `gpu-smoke`, the effective smoke surface in `tests/ops` is larger than needed for a representative operator-level signal. As a result, even when the workflow stays on the smoke tier, the test set is still too broad for a fast PR critical path.

### Root Cause Analysis
The `tests/ops` smoke tier had grown into a broad historical superset rather than a compact representative set. There was no single enforced rule ensuring that each public operator exported from `tileops/ops/__init__.py` contributes exactly one canonical smoke case to the PR path.

Without a centralized ownership rule, the effective smoke surface can expand over time as more tests are added, which directly undermines the P0 goal of shrinking `Run tests` time.

### Related Files
- `output/gpu-smoke-perf/README.md`
- `output/gpu-smoke-perf/todo.md`
- `tileops/ops/__init__.py`
- `tests/conftest.py`
- `tests/ops/`
- `.github/workflows/gpu-smoke.yml`

## Goal
Reduce the `tests/ops` smoke scope used by `gpu-smoke` for PR validation so that it becomes a compact, deterministic operator-representative set aligned with the P0 optimization target in `output/gpu-smoke-perf/README.md`.

## Plan

1. Define a canonical smoke ownership rule for `tests/ops`: every public `Op` exported from `tileops/ops/__init__.py` must map to exactly one smoke test entry.
2. Require canonical smoke entries to use `tune=False` / `autotune=False` when applicable.
3. Normalize non-canonical `tests/ops` cases to `full` so that smoke remains minimal and stable.
4. Fail collection when a new public `Op` is exported without a declared canonical smoke target.
5. Validate the resulting smoke set with collection-only checks and compare the selected smoke surface against the intended PR critical-path scope.

## Constraints
- Do not widen the smoke tier while implementing this reduction.
- Keep the existing smoke/full/nightly tier hierarchy intact.
- Preserve broad `full` coverage; this issue is about shrinking PR smoke, not reducing overall correctness coverage.
- Restrict the canonical smoke rule to `tests/ops`; non-operator structural tests can keep their own tier semantics.

## Acceptance Criteria
- [ ] Modified files pass unit tests
- [ ] PR-path `tests/ops` smoke coverage is reduced to a canonical operator-representative set
- [ ] Every public `Op` exported from `tileops/ops/__init__.py` has exactly one canonical smoke entry
- [ ] Canonical smoke entries use `tune=False` / `autotune=False`
- [ ] Adding a new public `Op` without updating the canonical smoke mapping fails collection with a clear error
- [ ] `python -m pytest tests --collect-only -q` passes
- [ ] `python -m pytest tests/ops --collect-only -q -m smoke` selects only the canonical smoke set

### PR DESCRIPTION
**Pull request by RMLYC**
Closes #736

## Summary

- centralize canonical smoke ownership for `tests/ops` in `tests/conftest.py` using a single `PUBLIC_OP_SMOKE_TARGETS` mapping keyed by public Ops from `tileops/ops/__init__.py`
- normalize non-canonical `tests/ops` smoke cases to `full` so PR `gpu-smoke` runs keep exactly one `smoke` case per public Op with `tune/autotune=False`

## Test plan

- [x] `python -m pytest tests --collect-only -q`
- [x] `python -m pytest tests/ops --collect-only -q -m smoke`

## Benchmark

- this change implements the P0 smoke-scope reduction in `output/gpu-smoke-perf/README.md`
- after normalization, `python -m pytest tests/ops --collect-only -q -m smoke` selects `80/1502` collected `tests/ops` cases for the PR critical path

## Additional context

- `tests/ops` canonical smoke assignment is now enforced at collection time, and new public Ops must register exactly one smoke target in `tests/conftest.py`

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -1,10 +1,304 @@
...
+# The collected smoke set is assigned dynamically during collection, so
+# tests/ops sources do not need to preserve static pytest.mark.smoke markers.
+# When a new public Op is exported from tileops/ops/__init__.py, add exactly
+# one canonical smoke target here and let collection normalize the remaining
+# tests/ops cases to full.
+PUBLIC_OP_SMOKE_TARGETS: dict[str, tuple[str, str]] = {
+    "AdaLayerNormOp": ("tests/ops/test_ada_layer_norm.py", "test_ada_layer_norm_op"),
+    "AdaLayerNormZeroOp": (
+        "tests/ops/test_ada_layer_norm_zero.py",
+        "test_ada_layer_norm_zero_op",
```
Comment (lcy-seso) [inline tests/conftest.py:27]: ~~**[Design]** Retracted — this suggestion contradicts the PR's centralization intent. If canonical targets keep source-level smoke markers, the mapping table becomes redundant. The decoupling is a deliberate design choice, not a problem.~~

File: tests/conftest.py [hunk 2]
```diff
@@ -1,10 +1,304 @@
...
+    "TopkSelectorOp": ("tests/ops/test_topk_selector.py", "test_topk_selector_op"),
+    "UnaryOp": ("tests/ops/test_activation.py", "test_relu_op"),
+    "VarMeanOp": ("tests/ops/test_reduce.py", "test_var_mean_op"),
+    "VarOp": ("tests/ops/test_reduce.py", "test_var_op"),
+}
+
+
+def _load_public_op_names() -> list[str]:
+    tree = ast.parse(OPS_INIT_PATH.read_text())
+    for node in tree.body:
```
Comment (lcy-seso) [inline tests/conftest.py:171]: ~~**[Design]** Retracted — observation about coupling is valid but secondary. The main concern with this PR is the smoke selection granularity (see follow-up comment), not the mapping mechanism.~~

File: tests/conftest.py [hunk 3]
```diff
@@ -1,10 +1,304 @@
...
+        if not rel_path.startswith(OPS_TEST_PREFIX):
+            continue
+        if item.nodeid in selected_nodeids:
+            _set_explicit_tier(item, "smoke")
+        elif item.get_closest_marker("smoke") is not None or all(item.get_closest_marker(name) is None for name in TIER_NAMES):
+            _set_explicit_tier(item, "full")
+
+    smoke_counts: dict[str, int] = defaultdict(int)
+    for op_name, item in selected_smoke_items.items():
+        if item.get_closest_marker("smoke") is None:
```
Comment (lcy-seso) [inline tests/conftest.py:288]: ~~**[Code]** Previous analysis was incorrect — retracted.~~

Correction: `_clear_inherited_tiers_for_ops_items` only strips tier markers from **parent nodes** (class/module), not from item `own_markers`. Items that had `smoke` in their own markers still retain it after clearing. The `get_closest_marker("smoke")" check correctly handles non-canonical items that still carry a source-level smoke marker. Not dead code.

File: tests/conftest.py [hunk 4]
```diff
@@ -1,10 +1,304 @@
...
+        rel_path = _relative_test_path(item.path)
+        if not rel_path.startswith(OPS_TEST_PREFIX):
+            continue
+        node = item.parent
+        while node is not None:
+            node_id = id(node)
+            if node_id not in visited_nodes:
+                node.own_markers = [m for m in node.own_markers if m.name not in TIER_NAMES]
+                visited_nodes.add(node_id)
+            node = node.parent
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tests/conftest.py]: Fixed in the latest update. `_clear_inherited_tiers_for_ops_items()` now tracks visited parent nodes by the node object itself instead of `id(node)`, which removes the unnecessary object-id reuse concern without changing the collection behavior.

File: tests/conftest.py [hunk 5]
```diff
@@ -1,10 +1,304 @@
...
+            if isinstance(target, ast.Name) and target.id == "__all__":
+                return sorted(
+                    elt.value
+                    for elt in node.value.elts
+                    if isinstance(elt, ast.Constant)
+                    and isinstance(elt.value, str)
+                    and elt.value.endswith("Op")
+                    and elt.value != "Op"
+                )
+    raise RuntimeError(f"Failed to load __all__ from {OPS_INIT_PATH}")
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tests/conftest.py:184]: I agree this is a tradeoff, but it is a deliberate one for this PR. I avoided importing `tileops.ops` during collection-time validation because that would couple smoke selection to module import side effects and environment readiness. The current `ast` read is intentionally a lightweight static read of `__all__`. I think it is acceptable for the P0 scope here, while a stronger source-of-truth design (for example an explicit manifest/config) is better handled as a separate follow-up change.

File: tests/conftest.py [hunk 6]
```diff
@@ -1,10 +1,304 @@
+import ast
 from collections import defaultdict
+from pathlib import Path
 
 import pytest
 import torch
 
 from tests.test_base import _check_result
 
+REPO_ROOT = Path(__file__).resolve().parent.parent
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tests/conftest.py:10]: This is a reasonable robustness concern, but I would prefer not to expand this PR into a broader repo-root discovery redesign. In this specific case the path is derived from __file__, so it is not sensitive to the caller working directory; the real coupling is to the current repository layout. I think that tradeoff is acceptable for this P0 change, and if we want a more formal root or source-of-truth mechanism, I would rather handle that in a follow-up issue or PR.
